### PR TITLE
Functionality for weighted graphs

### PR DESCRIPTION
**Pull request by jkinable**
- added setEdgeWeight(u,v, weight) to graph interface. I'm debating whether we want a getEdgeWeight(u,v) as well, since these functions are so common.
- added functionality to AsWeightedGraph to define the weight of an edge as a function. This is very convenient if computing the distance matrix is very expensive. In one of my projects the weights are based on a stochastic function. Precomputing a 1000x1000 distance matrix implies sampling 1000000 random distributions. When providing a weight function, the weight is only computed when it is actually needed.

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/graph/AsWeightedGraph.java [hunk 1]
```diff
@@ -47,8 +47,10 @@
 {
 
     private static final long serialVersionUID = -6838132233557L;
+    private final Function<E, Double> weightFunction;
     private final Map<E, Double> weights;
-    private final boolean writeWeightsThrough;
+    public final boolean writeWeightsThrough;
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/graph/AsWeightedGraph.java]: These two need to be private (possible with a getter).

File: jgrapht-core/src/main/java/org/jgrapht/Graph.java [hunk 2]
```diff
@@ -576,5 +576,21 @@
      */
     void setEdgeWeight(E e, double weight);
 
+    /**
+     * Assigns a weight to an edge between <code>sourceVertex</code> and <code>targetVertex</code>.
+     * If no edge exists between <code>sourceVertex</code> and <code>targetVertex</code> or either of these vertices is
+     * <code>null</code>, a <code>NullPointerException</code> is thrown.
+     * <p> When there exist multiple edges between <code>sourceVertex</code> and <code>targetVertex</code>, consider using
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/Graph.java]: Needs some more documentation. What happens if there are multiple edges between source and target. Do they all change weights or just the first one.

File: jgrapht-core/src/main/java/org/jgrapht/graph/AsWeightedGraph.java [hunk 3]
```diff
@@ -98,17 +129,27 @@ public AsWeightedGraph(Graph<V, E> graph, Map<E, Double> weights, boolean writeW
...
-        }
 
         return weight;
     }
 
     /**
-     * Assigns a weight to an edge.
+     * Assigns a weight to an edge. If <code>writeWeightsThrough</code> is set to <code>true</code>, the same weight is
+     * set in the backing graph. If this class was constructed using a weight function, it only makes sense to invoke this
+     * method when <code>cacheWeights</code> is set to true. This method can than be used to preset weights in the cache, or
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/graph/AsWeightedGraph.java]: "than"->"then"

File: jgrapht-core/src/main/java/org/jgrapht/graph/AsWeightedGraph.java [hunk 4]
```diff
@@ -98,17 +129,27 @@ public AsWeightedGraph(Graph<V, E> graph, Map<E, Double> weights, boolean writeW
...
+        if (Objects.isNull(weight))
             weight = super.getEdgeWeight(e);
-        }
 
         return weight;
     }
 
     /**
-     * Assigns a weight to an edge.
+     * Assigns a weight to an edge. If <code>writeWeightsThrough</code> is set to <code>true</code>, the same weight is
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/graph/AsWeightedGraph.java]: Maybe the explanation here needs to be expanded.

**Comment by jkinable** [comment]
@d-michail I've tried to update the description. Initially I just kept the original description, but this indeed gets rather confusing.
To simplify this class, I would personally be in favor of removing the writeThrough functionality. The user might as well invoke setEdgeWeight directly on the backing graph. What do you think?

**Comment by jsichi** [comment]
Note that as of the last release and all previous releases, the writeThrough behavior was always implicitly performed whenever the underlying graph was weighted (matching the current parameter default).  So I think it would be messy to remove it.

**Comment by jkinable** [comment]
Alright let's keep it this way then. Feel free to merge when satisfied.

**Comment by jkinable** [comment]
@d-michail if you are happy with this one, could you go ahead and merge? I'd like to get this one in before the release.

**Comment by d-michail** [comment]
Thanks!

**Pull request merged**